### ISSUE TITLE
HTTP/2 design considerations

### ISSUE THREAD

**Issue by bel2125** [civetweb/civetweb#875]
The HTTP2 standard is available for some time now, and the first request to support is on the TODO list since ~3 years. Unfortunately including a third party library (like suggested in #324) seemed to be not feasible. Even a library has a compatible license (MIT), it must also have a compatible architecture. Existing applications should work with little to no changes.
"No change" would be possible only with a HTTP1-to-HTTP2-proxy approach. There is no meaning to include this directly into CivetWeb, as there are already existing stand-alone proxies for this.
"Little changes" means, it should keep the existing structure with registering callbacks and sending the HTTP response messages there. The callback should not need to know, if the client uses HTTP/1.x or HTTP/2.
I did not find an existing solution for this, that could be smoothly integrated into the CivetWeb source. Thus, it was required to create several prototypes to find a way this could work.

In this issue I want to explain some design details and collect feedback.

**Comment by bel2125** [comment]
A first important design consideration is, HTTP/2 splits requests/responses into frames on the wire. A frame is basically a block of data with variable size sent using an existing TCP connection. They are ~10 different frame types, in particular "headers" and "body" of HTTP are two different frames.
Up to now, a handler function could send a complete response like
```C
int my_request_handler(struct mg_connection *conn, void *cbdata)
{
  mg_printf(conn,
        "HTTP/1.0 200 OK\r\n"            // Start response, status header
        "Content-Type: text/plain\r\n"   // More response header lines
        "Connection: close\r\n\r\n"      // Last response header line
        "This is the content\r\n");      // Response body

  while (random()) {
    mg_printf(conn, "Some extra content\r\n");  // Response body ... unknown end
  } 
  return 200;
}
```
For HTTP/1.x, this data is passed directly to the TCP connection:
```text
HTTP/1.0 200 OK
Content-Type: text/plain
Connection: close

This is the content
Some extra content
```
For HTTP/2, the data on the TCP connection is binary for the header, and remains plain text for the body. It must be sent like
```text
BINARY: Begin header for request ID (stream ID) 5 - this header will have 42 bytes (some number)
BINARY: Status=200
BINARY: Content-Type: text/plain
BINARY: Begin data for request ID 5 - with 21 bytes (strlen of following text)
This is the content
BINARY: Begin data for request ID 5 - with 20 bytes (strlen of following text)
Some extra content
BINARY: End data for request ID 5 - with 0 bytes (strlen of following text)
```
The "Connection: close" header must not be sent.


There are two options to implement this. Either CivetWeb has to buffer all data sent by `mg_printf`/`mg_write` until it detects the end of the header part. Then it has to interpret and re-encode this header. Possible, but not very effective - in particular not for the internal functions, e.g., for sending a file from the disk.
The other option would be to use new functions for headers. This look like
```C
int my_request_handler(struct mg_connection *conn, void *cbdata)
{
  mg_response_start(conn, 200);
  mg_response_add_header(conn,  "Content-Type", "text/plain", -1);
  mg_response_add_header(conn,  "Connection", "close", -1);  /* if the connection is HTTP/2, this header will not be sent */
  mg_response_send_headers(conn);

  mg_printf(conn,
        "This is the content\r\n");      // Response body

  while (random()) {
    mg_printf(conn, "Some extra content\r\n");  // Response body ... unknown end
  } 
  return 200;
}
```

This handler callback will work for HTTP/1.0, HTTP/1.1, HTTP/2 and (most likely) HTTP/3.
Internal handler functions will use these new functions.
For user defined handler callbacks, there are three options:
1) respond to HTTP/2 requests to user callbacks with a "this endpoint (URL) supports HTTP/1.1 only (`HTTP_1_1_REQUIRED`)". CivetWeb will only serve static files (HTML, JS, CSS, images, ...). This is a valid response. Firefox/Chrome/Edge will retry using HTTP/1.1, curl will return an error with code 'HTTP_1_1_REQUIRED (13)' even if -L (follow redirects) is set.
2) require all callbacks to use new `mg_response_*` functions, if HTTP/2 is enabled. If a handler does not use the new functions, a HTTP/2 client will receive invalid data and close the connection with a protocol error. 
3) buffer all data sent by `mg_printf`/`mg_write` until it detects the end of the header part. Then it has to interpret and re-encode this header - but only for user defined handler callbacks. 

1 and 2 could be combined as a configuration option. The option would be to 
a) enable HTTP2 for user defined handler callbacks (mg_request_handler) and internal functions (like send a file from disk)
b) enable HTTP2 for internal functions (send a file) but respond with a HTTP_1_1_REQUIRED if a user defined callback (or Lua script)
c) disable HTTP2 for all (even if HTTP2 was enabled at compile time)

**Comment by bel2125** [comment]
Since all (major) browsers and some libraries support HTTP/2 only for secure (TLS) connections (https), it makes little sense to implement HTTP/2 without TLS (http). CivetWeb will support HTTP/2 using [ALPN](https://en.wikipedia.org/wiki/Application-Layer_Protocol_Negotiation). There seems to be no requirement for HTTP/2 without ALPN.

**Comment by xtne6f** [comment]
I highly recommend that you once revert HTTP2-related commits from master branch, then rebase/re-create the HTTP2 branch. I have already found some bugs related to the commits, but I can not decide where I should create the pull request.
Before the problem gets complicated, I hope they are clealy separated.

**Comment by bel2125** [comment]
I tried to add only some features to the master branch, that are not "directly" HTTP2, but also separate features usable without HTTP2.
 
One feature added to the master branch is ALPN. This can be used without HTTP2 - although it might be of limited use there.

The second feature new in the master branch is HTTP response header buffering. Also this feature is not related to HTTP2, and can be used in HTTP1.x as well. It should also help with headers that are set multiple times - e.g. in a callback and in a framework.

Did you find a bug in one of these two features? They should be fixed in master.

Other features ... pure HTTP2 features ... are not merged into master right now.

**Comment by bel2125** [comment]
There is only a placeholder without any functionality in master:

```C
#ifdef USE_HTTP2
#include "mod_http2.inl"
#endif

#if defined(USE_HTTP2)
	process_new_http2_connection(conn);
#endif
```

And an mod_http2.inl file without any features:

https://github.com/civetweb/civetweb/blob/fe1eef775b3e2de4762658febe5ccb0570016ffd/src/mod_http2.inl#L1-L5

If `USE_HTTP2` is not defined, it should do nothing.

**Comment by xtne6f** [comment]
> One feature added to the master branch is ALPN. This can be used without HTTP2 - although it might be of limited use there.

Yes, but no merit for HTTP/1.1. I suggest enclosing the codes for ALPN by `#if defined(USE_HTTP2)`.
Specifically, if the server does not support ALPN (if ServerHello does not contain ALPN extension), the client will send HTTP/1.1 as before. There is no overhead. (Use a packet analyzer. `Upgrade: h2` will not be sent by modern clients.)
I found a bug on the DLL loader because of the modification for ALPN.

Are the codes enclosed by `#if defined(USE_HTTP2)` not the pure HTTP2 features?
As this issue title says, the features are incomplete. Therefore, the codes will be added many changes from now on.
To avoid troublesome duplicated commits and code conflicts, I suggest organizing the situation once. It is not so fun work (if you permit, I will create some reference commits), but I believe it is worth doing.

> Do you think it would be helpful, to replace mod_http2.inl with a #error "HTTP2 not supported yet"?

HTTP2 assumes "multi stream per thread" model, and such model must have some advantages for HTTP/1.1 too. So I expect that the separated function like `process_new_http2_connection` would be integrated after all.

**Comment by bel2125** [comment]
> I suggest enclosing the codes for ALPN by #if defined(USE_HTTP2).

From what I heard/read, ALPN and HTTP/2 are two distinct features. HTTP/2 requires ALPN.
There are other ALPN protocols (https://www.iana.org/assignments/tls-extensiontype-values/tls-extensiontype-values.xhtml#alpn-protocol-ids), but probably none of them makes sense inside civetweb. If any of them could ever make sense, a #define(USE_ALPN) might be better, and USE_HTTP could set USE_ALPN - or do you think ALPN will never be used for something else?

> (Use a packet analyzer. Upgrade: h2 will not be sent by modern clients.)

Sorry, I don't understand what you mean. 

Do you have a packet analyzer that will also work as a test client sending "Upgrade" requests?
From what I know, a "general" packet analyzer (WireShark) will not send anything, and I still don't have a specialized HTTP analyzers supporting HTTP/2 (https://www.telerik.com/forums/http-2-support).

Or did you mean "Browsers support HTTP/2 only through ALPN, not through Upgrade" (this is what I meant with https://github.com/civetweb/civetweb/issues/875#issuecomment-633691638).

> I found a bug on the DLL loader because of the modification for ALPN.

What is this bug? Do you know how to solve it?
Could you create a PR for master? You could also set the ALPN in #ifdef(USE_HTTP2) or #ifdef(USE_ALPN), if you like.

> Are the codes enclosed by #if defined(USE_HTTP2) not the pure HTTP2 features?

Yes they are (or at least they should be).

> As this issue title says, the features are incomplete. Therefore, the codes will be added many changes from now on.

I tried to prepare it so only mod_http2.inl needs to be replaced for a first working version.  This way, HTTP2 development is detached, and does not need to be done using git. There is just one commit at the end, to publish the code. But for this I needed to establish an interface / "extension point" first. Without this separation, I was afraid the final merge will have dozens of conflicts and might accidentally revert some of the work done by others in the meantime.

> To avoid troublesome duplicated commits and code conflicts, I suggest organizing the situation once. It is not so fun work (if you permit, I will create some reference commits), but I believe it is worth doing.

I tried to organize this situation by separating in mod_http2.inl and "others". At least, this was my intention, but maybe I made some mistakes.
How would you do it?

**Comment by xtne6f** [comment]
> From what I heard/read, ALPN and HTTP/2 are two distinct features. HTTP/2 requires ALPN.

Yes, I have a question about the advantage that HTTP/1.1-only (`USE_HTTP2` is undefined) server supports ALPN.

> Or did you mean "Browsers support HTTP/2 only through ALPN, not through Upgrade" (this is what I meant with #875 (comment)).

Yes, I just meant that.

> What is this bug? Do you know how to solve it?

The current `struct ssl_func` uses an union, its `required` field is overrided after the first initialization.
If the initialization failed by any "Missing function", the second initialization (e.g. the second `mg_init_library()`) would be inaccurate or undefined behavior.
This is not all, but it is roughly like this.

> But for this I needed to establish an interface / "extension point" first. Without this separation, I was afraid the final merge will have dozens of conflicts and might accidentally revert some of the work done by others in the meantime.

But can you really see through the final shape of the interface / "extension point"? At least, I can't.
I also afraid the conflicts, but they are avoidable by frequent merging from master to HTTP2 branch.

> I tried to organize this situation by separating in mod_http2.inl and "others". At least, this was my intention, but maybe I made some mistakes.
How would you do it?

I expect that HTTP2 would be a superset of HTTP/1.1 in terms of implementation.
In other words, HTTP/1.1 would be a functionally limited version (one stream per connection) of HTTP2.
I can not propose detailed design for now, and that's why I think the additions related to HTTP2 should be limited to the branch for now.

**Comment by bel2125** [comment]
> Yes, I have a question about the advantage that HTTP/1.1-only (USE_HTTP2 is undefined) server supports ALPN.

For HTTP1.0/1.1, there is none.
Probably there is/will be some other protocol negotiated using in ALPN, that might be interesting to be added to civetweb.

If not, there probably is no benefit in using it except for HTTP/2. 

I just checked the origins of the protocols again: both HTTP/2 and ALPN have been developed by Google, and they are intended to be used together. Probably nobody *really* uses it for something else, even if it could be used.

> The current struct ssl_func uses an union, its required field is overrided after the first initialization.
If the initialization failed by any "Missing function", the second initialization (e.g. the second mg_init_library()) would be inaccurate or undefined behavior.

Ok, I see. I did not expect `mg_init_library()` to be called twice - and I did not expect a failed initialization due to missing functions could succeed anytime later (you need to install TLS library to make this work).
But using `struct` instead of `union` is a simple fix.

> This is not all, but it is roughly like this.

You mean, there is another bug?

> But can you really see through the final shape of the interface / "extension point"? At least, I can't.

After some theoretic considerations and some tests that demonstrated what ideas will not work out, I can at least see is a first "minimum viable" version. This minimum will not have multiple parallel streams for a connection, but still handle them sequentially. And it will not support server pushes. But it will handle header compression, and pipe-lining. So, this minimum will not provide all benefits from HTTP/2, but some reasonable subset. And it can be extended later.
The features included in the first version are the ones that are "structurally most compatible" with civetweb architecture.
Once they are included, we can take care of the more complicated topics, like the one from your post before (https://github.com/civetweb/civetweb/issues/875#issuecomment-647441419):

> HTTP2 assumes "multi stream per thread" model, and such model must have some advantages for HTTP/1.1 too. So I expect that the separated function like process_new_http2_connection would be integrated after all.

HTTP2 assumes "multi parallel streams per connection" - the RFC does not care about threads, that's an implementation detail. Actually, I think this is one of the most important design considerations and decisions that must be made. Probably I should handle this topic in a separate comment.

> I also afraid the conflicts, but they are avoidable by frequent merging from master to HTTP2 branch.

My plan to avoid merge conflicts was to merge complete individual/isolated features. Like ALPN. Like the "minimum viable" described above.
Maybe this does not work out as I thought ...

> I expect that HTTP2 would be a superset of HTTP/1.1 in terms of implementation.
In other words, HTTP/1.1 would be a functionally limited version (one stream per connection) of HTTP2.

I think so too. Since HTTP/2 requires more resources at the server side as compared to HTTP/1.1, it will probably stay optional. Some (embedded) devices may not profit from HTTP/2 at all (although HTTP/3 might be interesting, and this shares some concepts with HTTP/2).

> I can not propose detailed design for now, and that's why I think the additions related to HTTP2 should be limited to the branch for now.

HTTP/2 is a huge standard (96 pages [RFC 7540](https://tools.ietf.org/html/rfc7540), 55 pages [RFC 7541], other required stuff like ALPN, ...). It took me quite a while to get some overview, not just reading some shallow "HTTP/2 is great" articles. 

There are some design choices that must be made to integrate HTTP/2. 
And that's why I opened this issue: To explain and discuss the design decisions, and get feedback if something might be a bad decision for some reason I did not think of in the first place.

**Comment by xtne6f** [comment]
> Probably there is/will be some other protocol negotiated using in ALPN, that might be interesting to be added to civetweb.

If so, it's not too late even after such a protocol has been developed.
You have the right to decide finally, so I do not complain further, but, personally, I don't like the increase of meaningless binary footprint.

> You mean, there is another bug?

Yes, but they are trivial things (`ebuf` in `load_tls_dll()` might overrun, and so on.). Anyway I'm going to send a pull request to your branch.
There are also bugs about the closing of the DLLs (https://github.com/xtne6f/civetweb/commit/a60dca3db20fcc513a9fd9c323337a0d40ed80d8), though it is another issue.

Umm... What I want to say is only this: why don't you move the codes enclosed by `#if defined(USE_HTTP2)` on the master branch into HTTP2 branch. Such incomplete codes should cause troublesome things later.
However, if you don't want, I don't care that you deny my suggestion.

**Comment by gajanak** [comment]
Hi, 
little bug found on experimenting with USE_HTTP2
in civetweb.c  handle_request line 14066
ri->local_uri_raw could be NULL - i dont know why ;)
but adding if (ri->local_uri_raw != NULL) { ... } around the local_uri generation works for me.

PS: Is already any more work on h2 - not in master, already ? 
And Thanks for this great work. It make fun to read and understand your code.

**Comment by bel2125** [comment]
Thank you for the patch.

**Comment by bel2125** [comment]
Regarding the current implementation status of HTTP/2:

The HTTP/2 implementation since CivetWeb V1.13 is understood by all browsers.
Like all experimental implementations, it has been there for a while to get feedback from different user groups.

A strong limitation is HTTP/2 is only available for HTTPS but not for HTTP. This is not a CivetWeb limitation, but a decission of browser manufacturers. 
While it is a good idea to enable HTTPS for a web server in the internet, it es EXTREMELY cumbersome for embedded web servers, because of the required certificate.
This includes using it as a webserver running on all kinds of devices that are not directly connected to the internet and extending existing applications with a web interface.
The only solution is becomming your own certificate authority - this can be done (also with civetweb), but usually requires a larger, professional IT department, and even they think twice before implementing the required certificate management processes.
This makes HTTP/2 effectively unusable for a large set of relevant usecases. Devices like WiFi routers, printers, (home) automation, ... use HTTP, not HTTPS (at least in the default configuration).
Furthermore, the problems solved by HTTP/2 are not the most relevant problems you are facing for embedded webservers anyway. 
In other words, the feedback from the embedded webserver uses: HTTP/2 is not that useful anyway.

The feedback from using CivetWeb with HTTP/2 in internet, or from those users who handled the certificate topic somehow:
Header compression is implemented, and works.
Server pushes would be relatively easy to implement, but nobody asked for it up to now. We would need some configuration "if someone asks for page /a/index.html, send /a/resources/*.* as server push".
If you have a server running in a LAN with a small number of known clients, the clients will have /a/resources/*.* in the cache pretty soon and don't load it from the server anymore.
If you have a server running in the internet, maybe you will have these resources in a CDN or use some combination of two webservers - I don't know all the exact reasons, I just know nobody asked for it up to now.
Multiplexing and prioritization is not that easy to implement, and full independence of streams cannot be reached in HTTP/2 anyway, due to the underlying TCP protocol (therefore, HTTP/3 uses UDP). But it can be done if it solves a real problem.

So the status of HTTP/2 is, I am collecting feedback what you want to use and why?
I don't want to implement something that will not be used by anyone, since it does not solve any ot the real problems of the CivetWeb users.

**Comment by gajanak** [comment]
The TLS is not THE problem, i think. Special in embedded, its a good way that libs like Yours does provide it. Thats a key-point while i start working with your lib. Using openssl ( I patched to use openssel V3 Interface and use clang for arm-target ;) works good), special for embeeded or IOT - there is a big requirement for low level libs like yours , IMHO.

HTTP/2 : I only test it for future use-cases, it - how you already wrote: - works with Browser most way."its usable" ;)
( i wonder if switching to 1.1 on any PUT is weired , but i think , this will change in future ;)

One big advantage for HTTP/2 is multiple requests performance, and I hope, in the future - replacing own websocket Protocol with advanced request/response in a long living connection.

Perhaps you read more from me in future ;)
Greets from Germany ;) (sorry for Bad englisch, by the way)
Thanks for the long answer.

**Comment by koukine** [comment]
We have found civetweb very useful in our new product (in development). Thank you guys.

The footprint is significantly smaller than nginx and it does the job well. 

However, we now require http/2.

I have built from the master's branch. http/2 data delivery is implemented only for static serving, which we don't use.

We use callbacks, but civet issues "require http 1.1" for those and most other types of queries.

Has any more thought been given to http/2 since your last comments?

We will try to tinker with the experimental implementation and see how far we get.

**Comment by bel2125** [comment]
CivetWeb can use HTTP/2 in callbacks as well. Just make sure you use the newer interface:
`mg_response_header_start/mg_response_header_add/mg_response_header_send` instead of just using `mg_printf` or `mg_write` for the header as well. The body can still be sent with `mg_printf` or `mg_write` for HTTP/2 but not the header.
It is also working in field tests.

Server-Push, multiplexing and prioritization is still not implemented and also not work in progress - still nobody explicitly requested these features. "Everything else" should work ("everything" = currently I don't have any other part of HTTP/2 in mind that does not work).

**Comment by koukine** [comment]
Thanks, @bel2125 for such a prompt response.

I cannot see how it works (I tried in master branch).

HTTP1_only macro is called for callbacks which forces "http2_must_use_http1".

When I commented HTTP1_only from the callback, it, expectedly, failed.

I see there is also a dedicated HTTP2 branch. I have not tried it. I have just checked and HTTP1_only is the same there.

**Comment by bel2125** [comment]
The HTTP2 branch is already merged into the master branch.
But you have to set the `USE_HTTP2` feature define for the master branch.

**Comment by koukine** [comment]
We do enable USE_HTTP2 and set enable_http2 to "yes".
The initial connection is HTTP2 and is handled by civetweb server.

However, civet's reply is to tell the client to use http1.

Are we looking at a different code?

Here is one of the lines which forces downgrade to http1:
https://github.com/civetweb/civetweb/blob/master/src/civetweb.c#L14946

```
	/* 7. check if there are request handlers for this uri */
	if (is_callback_resource) {
		HTTP1_only;
```

**Comment by bel2125** [comment]
I am afraid, yes, we are looking at different code.
In the field test I was using this line was removed. The master branch still has this compatibility assumption, that callbacks are not aware of HTTP/2 and are using `mg_printf` to write header lines as they did 10 years before. I could have added "USE_HTTP2_FOR_CALLBACKS", "USE_HTTP2_FOR_LUA", ... but instead I removed this line 
... to fix it "later", but "later" I forgot about it :-(
I will have a look at it and merge it to the master branch in the next days.

**Comment by koukine** [comment]
Thanks, @bel2125 
I did remove the line HTTP1_only and it failed.
But I will do more work with headers in our client and retest.
Also, DATA fragments are not supported (ie POSTs, etc). Is it correct?
